botocore/credentials.py

```diff
@@ -791,6 +820,206 @@ def _assume_role_base_kwargs(self, config):
+        raise NotImplemented()
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        """Returns SAML assertion when login succeeds, or None otherwise"""
+        raise NotImplemented()
+
+
+class SamlGenericFormsBasedAuthenticator(SamlAuthenticator):
+    username_field = 'username'
+    password_field = 'password'
+
+    def is_suitable(self, config):
+        return config.get('saml_authentication_type') == 'form'
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        verify = config.get('saml_verify_ssl') != 'false'
+        if not config.get('saml_username'):
+            config['saml_username'] = username_prompter("Username: ")
+        endpoint = config['saml_endpoint']
+        login_form = self._get_form(requests.get(endpoint, verify=verify).text)
+        if login_form is None:
+            raise SamlError(detail='Login form is not found in %s' % endpoint)
+        payload = dict((tag.attrib['name'], tag.attrib.get('value', ''))
+                       for tag in login_form.findall(".//input"))
+        if self.username_field in payload:
+            payload[self.username_field] = config['saml_username']
+        if self.password_field in payload:
+            payload[self.password_field] = password_prompter("Password: ")
+        response_form = self._get_form(requests.post(
+            urljoin(endpoint, login_form.attrib.get('action', '')),
+            data=payload, verify=verify).text)
+        if response_form is not None:
+            return self._get_value_of_first_tag(
+                response_form, 'input', 'name', 'SAMLResponse')
+        # Login failed, typically caused by incorrect username and/or password
+        return None
+
+    def _get_value_of_first_tag(self, root, tag, attr, trait):
+        ## This is backported from the following Python 2.7+ implementation:
+        # found = root.findall(".//tag[@attr='trait']")
+        # return found[0].attrib.get('value') if found else None
+        for element in root.findall(tag):
+            if element.attrib.get(attr) == trait:
+                return element.attrib.get('value')
+
+    def _get_form(self, html):
+        # Scrape a form from html page, and return it as an elementtree element
+        form_snippet = re.search('(<form.+</form>)', html, flags=re.DOTALL)
+        if form_snippet:
+            # To handle &nbsp;, on Python 2 we can use an undocumented parser:
+            #   ET.XMLParser().parser.UseForeignDTD(True)
+            # but it won't work on Python 3.
+            # So we use a pure XML way to handle it, for now.
+            return ET.fromstring(
+                '''<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
+                "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd" [
+                <!ENTITY nbsp ' '>
+                ]>''' + form_snippet.group(0))
+
+
+class SamlAdfsFormsBasedAuthenticator(SamlGenericFormsBasedAuthenticator):
```

Comment:
Does it also include changing the class `AssumeRoleWithSamlProvider` to `AssumeRoleWithSAMLProvider`? Such `ContainingCAPITALWordInTheMiddle` style just look a little bit weird to me.